### PR TITLE
JetStream Ordered consumer

### LINKED ISSUE #32: Ordered Consumer

### PR DESCRIPTION
**Pull request by mtmk**
Ordered consumer implementation based on https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-17.md

* Ordered consumer creation parameters
* Wrapper around the existing pull consume / fetch logic

**API change discussion**:

Simplified consumer api to just *consume*, *fetch* and *next* methods removing the `*All*` versions. Now *consume* and *fetch* methods (renamed from the `*All*` versions) return `IAsyncEnumerable` only. This reduces the API footprint as well as making it consistent between 'Ordered Consumer' and 'Consumer'.

* Removed consume and fetch methods exposing channel reader.
* Renamed `*All*` methods e.g. `FetchAllAsync()` to `FetchAsync()`.
* Also removed the interfaces exposing the channel readers.

resolves #32

### CODE REVIEW

**Comment by mtmk** [comment]
What can we do about the disparity with core subscribe?

Subscribe call on core on the other hand still returns an interface `INatsSub<T>` which exposes a channel reader for messages among a few others:
```csharp
public interface INatsSub<T> : IAsyncDisposable
{
    ChannelReader<NatsMsg<T>> Msgs { get; }
    string Subject { get; }
    string? QueueGroup { get; }
    public ValueTask UnsubscribeAsync();
}
```

An idea here maybe to introduce a different call to match JetStream calls:

```csharp
// New matching call
IAsyncEnumerable<NatsMsg<T>> SubscribeAsync<T>(string subject, ...);

// Existing call renamed
ValueTask<INatsSub<T>> SubscribeCoreAsync<T>(string subject, ...);
```

File: src/NATS.Client.JetStream/NatsJSContext.Consumers.cs [hunk 1]
```diff
@@ -125,4 +143,63 @@ public async ValueTask<bool> DeleteConsumerAsync(string stream, string consumer,
...
+        {
+            request.Config.FilterSubjects = opts.FilterSubjects;
+        }
+
+        var subject = $"{Opts.Prefix}.CONSUMER.CREATE.{request.StreamName}";
+
+        if (!string.IsNullOrWhiteSpace(request.Config.Name))
+        {
+            subject += $".{request.Config.Name}";
+            request.Config.Name = default!;
```
Comment (caleblloyd) [inline src/NATS.Client.JetStream/NatsJSContext.Consumers.cs:192]: why mutate name here?

File: src/NATS.Client.JetStream/NatsJSContext.Consumers.cs [hunk 2]
```diff
@@ -125,4 +143,63 @@ public async ValueTask<bool> DeleteConsumerAsync(string stream, string consumer,
...
+        }
+
+        if (!string.IsNullOrWhiteSpace(request.Config.FilterSubject))
+        {
+            subject += $".{request.Config.FilterSubject}";
+        }
+
```
Comment (caleblloyd) [inline src/NATS.Client.JetStream/NatsJSContext.Consumers.cs:199]: What if the name was not set, but the filter subject was set?  With this logic the subject token for name is not present, and the filter subject token is appended to where the name subject should go

File: src/NATS.Client.JetStream/NatsJSContext.Consumers.cs (hunk 1 above)
Comment (mtmk) [reply src/NATS.Client.JetStream/NatsJSContext.Consumers.cs:192]: spot on! no need for it.

**Review by caleblloyd — APPROVED** [review]
LGTM

**Pull request merged**